oktapam/data_source_group_test.go

```diff
@@ -4,119 +4,58 @@ import (
-	return func(s *terraform.State) error {
-		rs, ok := s.RootModule().Resources[rn]
-		if !ok {
-			return fmt.Errorf("resource not found: %s", rn)
-		}
-		mappings, err := getIndexMappingFromResource(rs, attributes.Groups, attributes.Name, len(expectedGroups))
-		if err != nil {
-			return fmt.Errorf("error mapping resources to indices: %w", err)
-		}
-
-		primaryAttributes := rs.Primary.Attributes
-		for name, group := range expectedGroups {
-			idx, ok := mappings[name]
-			if !ok {
-				return fmt.Errorf("could not find resource with %s: %s", attributes.Name, name)
-			}
-
-			name, ok := primaryAttributes[fmt.Sprintf("%s.%s.%s", attributes.Groups, idx, attributes.Name)]
-			if !ok {
-				return fmt.Errorf("%s attribute not set for group %q", attributes.Name, name)
-			}
-			if name != *group.Name {
-				return fmt.Errorf("%s attributes did not match for group %q, expected %q, got %q", attributes.Name, name, *group.Name, name)
-			}
-
-			expectedNumRoles := len(group.Roles)
-			numRoles, ok := primaryAttributes[fmt.Sprintf("%s.%s.%s.#", attributes.Groups, idx, attributes.Roles)]
-			if !ok {
-				return fmt.Errorf("%s attribute not set for group %q", attributes.Roles, name)
-			}
-			if numRoles != fmt.Sprint(expectedNumRoles) {
-				return fmt.Errorf("mismatch of number of %s set for group %q, expected %d, got %s", attributes.Roles, name, expectedNumRoles, numRoles)
-			}
-
-			roles, err := getArrayFromResource(rs, fmt.Sprintf("%s.%s.%s", attributes.Groups, idx, attributes.Roles), expectedNumRoles)
-			if err != nil {
-				return fmt.Errorf("error while retrieving roles from state: %w", err)
-			}
-
-			for _, role := range group.Roles {
-				if !arrayContainsString(roles, role) {
-					return fmt.Errorf("expected to have role %q set for group %q", role, name)
-				}
-			}
+// NOTE: This config (1) creates two new resources (2) lists the existing resources with the matching identifier
+// and (3) get the new resource as a data source.
+// The test then compares the resource with its data source to ensure they are equal.
 
-		}
-
-		return nil
-	}
+const testAccDatasourceGroupInitListFetchConfigFormat = `
+resource "oktapam_group" "test-1" {
+	name = "%s-1"
+	roles = ["access_user"]
 }
 
-const testAccDatasourceGroupsConfigFormat = `
-data "oktapam_group" "test_groups" {
-	contains = "%s"
+resource "oktapam_group" "test-2" {
+	name = "%s-2"
+	roles = ["access_user"]
 }
-`
 
-func createTestAccDatasourceGroupsConfig(prefix string) string {
-	return fmt.Sprintf(testAccDatasourceGroupsConfigFormat, prefix)
+data "oktapam_groups" "list" {
+	depends_on = [oktapam_group.test-1]
+	contains = "%s"
 }
 
-const testAccDatasourceGroupInitConfigFormat = `
-resource "oktapam_group" "test-group-1" {
-	name = "%s-1"
-	roles = []
-}
-resource "oktapam_group" "test-group-2" {
-	name = "%s-2"
-	roles = ["access_user"]
+data "oktapam_group" "target" {
```

Comment:
Not sure why do you need data "oktapam_project_groups" too in the test. You can just get the group name from the resource itself.

Reply:
I liked the idea of testing both in conjunction.

Reply:
One more thing, on naming I think better to put "test_acc" as prefix for consistency. Ideally all resources will be deleted within the lifecycle of a test, but there can be some situations arise where resource created are not deleted. Consistent naming will help in cleaning up.

Reply:
I think later on we will add some common code to get the test resource name based on resource type.

Reply:
I have some testing / example ideas in the works.